File: src/product-except-index/product_except_index.py
def get_products_of_all_ints_except_at_index( arr ):
	"""
	**********************************************************************************************************************
	You have an array of integers, and for each index you want to find the product of every
	integer except the integer at that index.

	Do not use division in your solution.
	"""
	length = len( arr )
	productArray = [1] * length
	product = 1
	
	for i in range ( length ):
		productArray[i] = product
		product *= arr[i]

	product = 1
	for i in range ( length - 1, 0, -1 ):
		productArray[i] *= product
		product *= arr[i]

	productArray[0] *= product
	return productArray


print( get_products_of_all_ints_except_at_index( [1, 7, 3, 4] ) )
# expected: [84, 12, 28, 21]

File: product-except-index/product-except-index.js
/**********************************************************************************************************************
 * You have an array of integers, and for each index you want to find the product of every
 * integer except the integer at that index.
 *
 * Do not use division in your solution.
 */
 function getProductsOfAllIntsExceptAtIndex( arr )
 {
 	// copy input array
 	var productArray = arr.slice();
 	// make two passes over array: front to back and back to front
 	var len = arr.length;
    var productSoFar = 1;
 	for ( var i = 0; i < len; i++ )
 	{
 		productArray[i] = productSoFar;
        productSoFar *= arr[i];
 	}
   	productSoFar = 1;
 	for ( var i = len - 1; i >= 0; i-- )
 	{
 		productArray[i] *= productSoFar;
        productSoFar *= arr[i];
 	}
   return productArray;
 }